File: bio-informatics/challenges/compute-frequencies.js
const patternToNumber = require('./pattern-to-number');

const computeFrequencies = (DNAString, kMerLength) => {
    const frequencyArray = new Array(4 ** kMerLength).fill(0);

    for (let start = 0; start < DNAString.length; start++) {
        const end = start + kMerLength;
        if(!DNAString[end -  1]) { break }

        const currentSequence = DNAString.slice(start, end);
        
        const index = patternToNumber(currentSequence);
        frequencyArray[index]++;
    }
    
    return frequencyArray;
}

module.exports = computeFrequencies;

File: bio-informatics/challenges/python/utils/compute_frequencies.py
import numpy as np
from utils import pattern_to_number

def compute_frequencies(dna_string, k_mer_length):
  frequency_array = np.zeros( (4 ** k_mer_length), dtype=int)

  for start in range(len(dna_string)):
    try:
      dna_string[(start + k_mer_length) - 1]
    except:
      break

    end = start + k_mer_length
    current_sequence = dna_string[start:end]
    index = pattern_to_number(current_sequence)
    frequency_array[index] += 1

  return frequency_array
